### PR TITLE
Default T in JSON to improve ergonomics of returning arbitrary JSON

### PR DESCRIPTION
**Pull request by mehcode**
This ...

```rust
use std::collections::HashMap;

use rocket_contrib::json::{JSON, Value};

type SimpleMap = HashMap<&'static str, Value>;

#[derive(Debug, Deserialize)]
struct UserForm {
    username: String,
    email: String,
    password: String,
}

#[post("/user", data = "<body>")]
fn post_user(body: JSON<UserForm>) -> JSON<SimpleMap> {
    println!("{:?}", body);

    JSON(map!(
        "status" => "ok",
    ))
}
```

Becomes ...

```rust
use rocket_contrib::json::JSON;

#[derive(Debug, Deserialize)]
struct UserForm {
    username: String,
    email: String,
    password: String,
}

#[post("/user", data = "<body>")]
fn post_user(body: JSON<UserForm>) -> JSON {
    println!("{:?}", body);

    JSON(map!(
        "status" => "ok",
    ))
}
```

Thoughts?

### CODE REVIEW

**Comment by impowski** [comment]
Neat.

**Comment by SergioBenitez** [comment]
I actually just got rid of the `map!` macro in favor of `json!` in https://github.com/SergioBenitez/Rocket/commit/5fabb43a1b62e00c1fdeecd6ed8e5d057be83ff1. Now you simply need to return `JSON<Value>`. Do you think this change is still worth it?

**Comment by mehcode** [comment]
Yes. I still think so. It's not as big as improvement as before but it still reads like an improvement.

---

Hmm.. I like how it nicely supports arrays now. But `JSON(json!( ... ))` reads a bit funny. I was thinking (before now) of adding a `json!` macro that expands to `JSON(map!( ... ))` but I can see why you'd want to keep the consistency of the responder wrapper.

Perhaps it can change again to `o!(...)` (or something) that is content-agnostic so you could do `MsgPack(o!(...))` for instance?

**Comment by SergioBenitez** [comment]
That means we'd have to rewrite the macro. Currently, we just use the implementation from `serde` directly. It's really nice not to have to maintain it.

We could make `json!` expand to a `JSON<Value>` directly, however.

**Comment by mehcode** [comment]
> That means we'd have to rewrite the macro. Currently, we just use the implementation from serde directly. It's really nice not to have to maintain it.

Why? Sorry if I wasn't clear. I just meant rename it. Because it just returns a thing that is deserializable shouldn't in theory work anywhere you work arbitrarily with deserializables? Or do I not understand how this is done internally?

**Comment by SergioBenitez** [comment]
> Because it just returns a thing that is deserializable [...]

It's returns a thing that's `Serialize`, not `Deserialize`. Anyway, I see what you're saying. Let's revisit this when there's more than one serialization format in contrib.

**Comment by mehcode** [comment]
> It's returns a thing that's Serialize, not Deserialize. Anyway, I see what you're saying. Let's revisit this when there's more than one serialization format in contrib.

Heh. Yeah I meant `Serialize`.

---

@SergioBenitez Fixed this branch.

**Comment by SergioBenitez** [comment]
Why isn't Travis running this? :\

**Comment by SergioBenitez** [comment]
Alright, I've come around on this.

Can you modify the docs and example to inform users about the default parameter? In particular, the `json!` macro code example should just return `JSON` and mention why that's okay.

File: contrib/src/json/mod.rs [hunk 1]
```diff
@@ -129,20 +129,23 @@ impl<T> DerefMut for JSON<T> {
 /// ```
 ///
 /// The return type of a macro invocation is
-/// [Value](/rocket_contrib/enum.Value.html). A value created with this macro
+/// [Value](/rocket_contrib/enum.Value.html), and is the default type parameter 
+/// of the JSON type. A value created with this macro
```
Comment (SergioBenitez) [inline contrib/src/json/mod.rs]: This needs to be word-wrapped at 80 columns.

Comment (SergioBenitez) [inline contrib/src/json/mod.rs]: Also, I'd like a slightly longer explanation. Kill the `,` and add another sentence explaining that it's the default value and as such you can return `JSON` without specifying the type.

**Comment by SergioBenitez** [comment]
Merged in 5086f0eb01bcd2a789993ab0444dbd14ede272d0.

**Pull request closed without merging**